motionplan/check.go

```diff
@@ -0,0 +1,430 @@
+//go:build !no_cgo
+
+// Package motionplan is a motion planning library.
+package motionplan
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/motionplan/ik"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+)
+
+// CheckPlan checks if obstacles intersect the trajectory of the frame following the plan. If one is
+// detected, the interpolated position of the rover when a collision is detected is returned along
+// with an error with additional collision details.
+func CheckPlan(
+	checkFrame referenceframe.Frame, // TODO(RSDK-7421): remove this
+	executionState ExecutionState,
+	worldState *referenceframe.WorldState,
+	fs referenceframe.FrameSystem,
+	lookAheadDistanceMM float64,
+	logger logging.Logger,
+) error {
+	plan := executionState.Plan()
+	startingInputs := plan.Trajectory()[0]
+	wayPointIdx := executionState.Index()
+
+	// ensure that we can actually perform the check
+	if len(plan.Path()) < 1 {
+		return errors.New("plan must have at least one element")
+	}
+	if len(plan.Path()) <= wayPointIdx || wayPointIdx < 0 {
+		return errors.New("wayPointIdx outside of plan bounds")
+	}
+
+	// construct solverFrame
+	// Note that this requires all frames which move as part of the plan, to have an
+	// entry in the very first plan waypoint
+	sf, err := newSolverFrame(fs, checkFrame.Name(), referenceframe.World, startingInputs)
+	if err != nil {
+		return err
+	}
+	// construct planager
+	sfPlanner, err := newPlanManager(sf, logger, defaultRandomSeed)
+	if err != nil {
+		return err
+	}
+	// This should be done for any plan whose configurations are specified in relative terms rather than absolute ones.
+	// Currently this is only TP-space, so we check if the PTG length is >0.
+	// The solver frame will have had its PTGs filled in the newPlanManager() call, if applicable.
+	if sfPlanner.useTPspace {
+		return checkPlanRelative(checkFrame, executionState, worldState, fs, lookAheadDistanceMM, sfPlanner)
+	}
+	return checkPlanAbsolute(checkFrame, executionState, worldState, fs, lookAheadDistanceMM, sfPlanner)
+}
+
+func checkPlanRelative(
+	checkFrame referenceframe.Frame, // TODO(RSDK-7421): remove this
+	executionState ExecutionState,
+	worldState *referenceframe.WorldState,
+	fs referenceframe.FrameSystem,
+	lookAheadDistanceMM float64,
+	sfPlanner *planManager,
+) error {
+	plan := executionState.Plan()
+	startingInputs := plan.Trajectory()[0]
+	currentInputs := executionState.CurrentInputs()
```

Comment:
should we be worried about a case where this gives us an empty slice?

Reply:
Yes, will validate.

Reply:
This should actually be a non-issue. If the current inputs (nil or not) are not valid, that will be caught when we try to `Transform` them.

Reply:
hmm, I don't think that's what we want here.
In `move_request.go` when we check for collisions with transient obstacles using `obstaclesIntersectPlan`, what informs us about a collision is the existence of an error. We do not care what the error is, just that there is one. 
I can then see how we might be stuck in an infinite loop if we are always errors on a `Transform`.

Reply:
> In move_request.go when we check for collisions with transient obstacles using obstaclesIntersectPlan, what informs us about a collision is the existence of an error. We do not care what the error is, just that there is one.

Perhaps this is what should be fixed then :)